File: app/src/main/java/com/metahumanz/pacilread/BookshelfFilter.java
package com.metahumanz.pacilread;

import com.metahumanz.pacilread.model.BookRecord;

import java.util.Locale;

public final class BookshelfFilter {
    private BookshelfFilter() {
    }

    public static boolean matches(BookRecord book, String query, String tag, String series, String status) {
        if (book == null) return false;
        String normalizedQuery = normalize(query);
        boolean queryMatch = normalizedQuery.isEmpty()
                || normalize(book.title).contains(normalizedQuery)
                || normalize(book.author).contains(normalizedQuery);
        boolean tagMatch = isEmpty(tag) || (book.tags != null && book.tags.contains(tag));
        boolean seriesMatch = isEmpty(series) || safe(book.series).equals(series);
        boolean statusMatch = isEmpty(status) || safe(book.readingStatus).equals(status);
        return queryMatch && tagMatch && seriesMatch && statusMatch;
    }

    private static String normalize(String value) {
        return safe(value).trim().toLowerCase(Locale.ROOT);
    }

    private static boolean isEmpty(String value) {
        return value == null || value.isEmpty();
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/BookshelfFilter.kt
package com.metahumanz.pacilread

import com.metahumanz.pacilread.model.BookRecord
import java.util.Locale

object BookshelfFilter {
    @JvmStatic
    fun matches(book: BookRecord?, query: String?, tag: String?, series: String?, status: String?): Boolean {
        if (book == null) return false
        val normalizedQuery = normalize(query)
        val queryMatch = normalizedQuery.isEmpty() ||
            normalize(book.title).contains(normalizedQuery) ||
            normalize(book.author).contains(normalizedQuery)
        val tagMatch = tag.isNullOrEmpty() || book.tags?.contains(tag) == true
        val seriesMatch = series.isNullOrEmpty() || safe(book.series) == series
        val statusMatch = status.isNullOrEmpty() || safe(book.readingStatus) == status
        return queryMatch && tagMatch && seriesMatch && statusMatch
    }

    private fun normalize(value: String?): String = safe(value).trim().lowercase(Locale.ROOT)

    private fun safe(value: String?): String = value ?: ""
}
